File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.ArrayDeque;
import java.util.Queue;

class Solution {
    private int[] dx = {0, 0, 1, -1};
    private int[] dy = {1, -1, 0, 0};

    public void bfs(int[][] maps) {
        int xEnd = maps.length;
        int yEnd = maps[0].length;
        Queue<int[]> queue = new ArrayDeque<>();
        queue.offer(new int[]{0, 0});

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int x = current[0];
            int y = current[1];

            for (int i = 0; i < 4; i++) {
                int xx = x + dx[i];
                int yy = y + dy[i];

                if (xx >= 0 && xx < xEnd && yy >= 0 && yy < yEnd) {
                    if (maps[xx][yy] == 1) {
                        maps[xx][yy] = maps[x][y] + 1;
                        queue.offer(new int[]{xx, yy});
                    }
                }
            }
        }
    }

    public int solution(int[][] maps) {
        int xEnd = maps.length;
        int yEnd = maps[0].length;
        bfs(maps);
        return (maps[xEnd - 1][yEnd - 1] != 1) ? maps[xEnd - 1][yEnd - 1] : -1;
    }
}

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const answer = 0;
    const xend = maps.length;
    const yend = maps[0].length;
    const dx = [0,0,1,-1];
    const dy = [1,-1,0,0];
    
    const bfs = () => {
        const queue = [[0,0]]; 
        while(queue.length) {
            const [x, y] = queue.shift();
            for (let i = 0; i < 4; i++) {
                const xx = x + dx[i];
                const yy = y + dy[i];
                if (xx>=0 && xx<xend && yy>=0 && yy<yend) {
                    if (maps[xx][yy] == 1) {
                        maps[xx][yy] = maps[x][y]+ 1;
                        queue.push([xx, yy]);
                    }
                }
                
            }
            
        }
        
        
    };
    bfs();
    if(maps[xend-1][yend-1] == 1) return -1;
    else return maps[xend-1][yend-1];
    
}